reinforcement_learning/rlclientlib/logger/event_batcher.h

```diff
@@ -0,0 +1,87 @@
+#include "event_queue.h"
+#include "ranking_event.h"
+#include "generated/RankingEvent_generated.h"
+#include "generated/OutcomeEvent_generated.h"
+
+namespace reinforcement_learning {
+  namespace utility {
+    template<typename TEvent>
+    class event_batcher {
+    public:
+      event_batcher() {};
+      void batch_serialize(data_buffer& oss, size_t& remaining, event_queue<TEvent>& queue, size_t _send_high_water_mark, size_t batch_size);
+    };
+
+    template<typename TEvent>
+    void event_batcher<TEvent>::batch_serialize(data_buffer& oss, size_t& remaining, event_queue<TEvent>& queue, size_t _send_high_water_mark, size_t batch_size) {
+
+      if (batch_size <= 0) {
+        throw("event_batcher::batch_serialize batch_size is 0 - message is too large.");
+      }
+
+      if (std::is_same<TEvent, ranking_event>::value) {
```

Comment:
It doesn't seem like a proper way to customize behavior based on type. It has to be either:
- Hidden behind some common interface within ranking_event/interaction_event implementations
or
- using template specialization: https://en.cppreference.com/w/cpp/language/template_specialization

Reply:
Also, these two branches look like exact copy paste of each other (except some type names, that can be 100% unified using templates)

Reply:
I was not able to find a good so we can simplify the code. But I will check your specialization here or we can discuss later.

---
In reply to: [228199517](https://github.com/VowpalWabbit/vowpal_wabbit/pull/1653#discussion_r228199517) [](ancestors = 228199517)

Reply:
Same as above.

---
In reply to: [228204171](https://github.com/VowpalWabbit/vowpal_wabbit/pull/1653#discussion_r228204171) [](ancestors = 228204171)

Reply:
It looks like you do not event need specialization here - just single template method (with probably some type traits tricks) should work